src/storage/src/hummock/sstable/delete_range_aggregator.rs

```diff
@@ -0,0 +1,328 @@
+            if !largest_user_key.is_empty()
+                && tombstone.start_user_key.as_slice().ge(largest_user_key)
+            {
+                continue;
+            }
+
+            if !smallest_user_key.is_empty()
+                && tombstone.end_user_key.as_slice().le(smallest_user_key)
+            {
+                continue;
+            }
+
+            if self.gc_delete_keys && tombstone.sequence <= self.watermark {
+                continue;
+            }
+
+            let start_key = if smallest_user_key.is_empty()
+                || tombstone.start_user_key.as_slice().gt(smallest_user_key)
+            {
+                key_with_epoch(tombstone.start_user_key.clone(), tombstone.sequence)
+            } else {
+                key_with_epoch(smallest_user_key.to_vec(), tombstone.sequence)
+            };
+            let end_key = if largest_user_key.is_empty()
+                || tombstone.end_user_key.as_slice().lt(largest_user_key)
+            {
+                tombstone.end_user_key.clone()
+            } else {
+                largest_user_key.to_vec()
+            };
+            delete_ranges.push((start_key, end_key));
+        }
+        delete_ranges
+    }
+}
+
+pub trait DeleteRangeIterator {
+    fn start_user_key(&self) -> &[u8];
+    fn end_user_key(&self) -> &[u8];
+    fn current_epoch(&self) -> HummockEpoch;
+    fn next(&mut self);
+    fn seek_to_first(&mut self);
+    fn seek(&mut self, target_key: &[u8]);
+    fn valid(&self) -> bool;
+}
+
+pub struct SingleDeleteRangeIterator {
+    agg: Arc<DeleteRangeAggregator>,
+    seek_idx: usize,
+}
+
+impl DeleteRangeIterator for SingleDeleteRangeIterator {
+    fn start_user_key(&self) -> &[u8] {
+        &self.agg.delete_tombstones[self.seek_idx].start_user_key
+    }
+
+    fn end_user_key(&self) -> &[u8] {
+        &self.agg.delete_tombstones[self.seek_idx].end_user_key
+    }
+
+    fn current_epoch(&self) -> HummockEpoch {
+        self.agg.delete_tombstones[self.seek_idx].sequence
+    }
+
+    fn next(&mut self) {
+        self.seek_idx += 1;
+    }
+
+    fn seek_to_first(&mut self) {
+        self.seek_idx = 0;
+    }
+
+    fn seek(&mut self, target_key: &[u8]) {
+        self.seek_idx = 0;
+        while self.seek_idx < self.agg.delete_tombstones.len()
+            && self.agg.delete_tombstones[self.seek_idx]
+                .end_user_key
+                .as_slice()
+                .le(target_key)
+        {
+            self.seek_idx += 1;
+        }
+    }
+
+    fn valid(&self) -> bool {
+        self.seek_idx < self.agg.delete_tombstones.len()
+    }
+}
+
+pub struct DeleteRangeAggregatorIterator<I: DeleteRangeIterator> {
+    inner: I,
+    end_user_key_index: BinaryHeap<DeleteRangeTombstone>,
+    epoch_index: BTreeSet<HummockEpoch>,
+    watermark: u64,
+}
+
+impl<I: DeleteRangeIterator> DeleteRangeAggregatorIterator<I> {
+    pub fn should_delete(&mut self, target_key: &[u8], epoch: HummockEpoch) -> bool {
```

Comment:
this function is correct
I have just checked equal case in this function, and it seems also correct.
You can check equal case in other codes and difference of user key and full key in other codes in you spare time.